x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ModelMeta.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.inference;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
+import org.elasticsearch.xpack.core.ml.utils.ToXContentParams;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class ModelMeta implements ToXContentObject, Writeable {
+
+    private static String MODEL_META = "model_meta";
+    private static ParseField MODEL = new ParseField("model");
+    private static ParseField TYPE = new ParseField("type");
+
+    public static final ConstructingObjectParser<ModelMeta, Void> STRICT_PARSER = createParser(false);
+    public static final ConstructingObjectParser<ModelMeta, Void> LENIENT_PARSER = createParser(true);
+
+    private static ConstructingObjectParser<ModelMeta, Void> createParser(boolean lenient) {
+        ConstructingObjectParser<ModelMeta, Void> parser = new ConstructingObjectParser<>("model_meta", lenient,
+                a -> new ModelMeta((String) a[0], (String) a[1]));
+
+        parser.declareString(ConstructingObjectParser.constructorArg(),  MODEL);
+        parser.declareString(ConstructingObjectParser.optionalConstructorArg(), TYPE);
```

Comment:
I think we will always need a `type`, whether it is top level or not, that is up for debate.

Additionally, are we wanting to store the model definition here as well? If it may make sense for `namedWritables` here as various models will have vastly different parameters, pre-processing, etc.